base/cvd/cuttlefish/host/libs/web/credential_source.h

```diff
@@ -29,6 +30,13 @@ class CredentialSource {
   virtual Result<std::string> Credential() = 0;
 };
 
+enum class CredentialSourceId {
+    kAcloud,
+    kLuci,
+    kLuciStorage,
+};
+using CredentialSources = std::unordered_map<CredentialSourceId, std::unique_ptr<CredentialSource>>;
```

Comment:
I don't see the point of this map yet. `fetch_cvd.cc` needs to know how to create each `CredentialSource` and how to instantiate each `BuildApi`, it could just as well pass the corresponding `CredentialSource` instance to each `BuildApi` constructor|factory. Am I missing something?

Reply:
I have MyBuildApi which inherits BuildApi, and CachingBuildApi<MyBuildApi>.  MyBuildApi needs multiple credentials.  To keep CachingBuildApi agnostic to the credential count, fetch_cvd passes in the map of credentials to CachingBuildApi, which just hands it off to whatever class it's subclassing (either BuildApi or MyBuildApi).

Reply:
That makes sense.

Are you planning to contribute the implementation of `MyBuildApi` to this repository? Without someone might change this in the future and break it for you downstream.

Reply:
Also, please consider passing a `std::unique_ptr<BuildApi>` to the constructor of `CachingBuildApi`. That would eliminate the need to pass the entire set of credential sources to `CachingBuildApi` and the need to use templates.

Reply:
Sure, lemme refactor.

The current MyBuildApi implementation is very Motorola-specific, so I hadn't planned on upstreaming it.  I'll re-review after I complete this current pull request, and see if there's parts that would make sense or be helpful to upstream.

Reply:
Ok, I refactored, and it's quite a bit cleaner.  bazel test pass, and no leaks are reported w/ valgrind.